### PR TITLE
FAI-182: Score Cards: Add skeleton for PMML Score Cards (4 tabs in UX design)

### PR DESCRIPTION
**Pull request by manstis**
Hi @tiagobento this is my first hack at adding a TypeScript (PMML) editor to `kogito-tooling` channels (_online_ only, atm)

It's just the last commit that's important (the others relate to other JIRA/PR that this is built on top of.. they're unmerged atm).

Could you please give me your opinion?

I also need to know how to integrate `StateControl` with the mock/stub/test `PMMLEditor`, if you can advise?

### CODE REVIEW

File: packages/embedded-editor/src/envelope/envelope.ts [hunk 1]
```diff
@@ -14,22 +14,29 @@
...
 import { DefaultXmlFormatter, GwtAppFormerApi, GwtEditorWrapperFactory } from "@kogito-tooling/kie-bc-editors";
 import * as MicroEditorEnvelope from "@kogito-tooling/microeditor-envelope";
+import { CompositeEditorFactory, EditorFactory } from "@kogito-tooling/microeditor-envelope";
 import { EnvelopeBusMessage } from "@kogito-tooling/microeditor-envelope-protocol";
+import { PMMLEditorFactory } from "@kogito-tooling/pmml-editor";
 
 const gwtAppFormerApi = new GwtAppFormerApi();
 gwtAppFormerApi.setClientSideOnly(true);
 
+const editorFactories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
```
Comment (manstis) [inline packages/embedded-editor/src/envelope/envelope.ts]: TypeScript (or possibly `TSLint`) complains bitterly about using `[]` for _typed_ types (i.e. `EditorFactory<LanguageData>`) and use of `Array` is forced...

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  public supports(languageData: LanguageData) {
+    return true;
+  }
+
+  public createEditor(
+    languageData: LanguageData,
+    messageBus: EnvelopeBusInnerMessageHandler
+  ): Promise<AppFormer.Editor> {
+    const candidates: Array<EditorFactory<LanguageData>> = this.factories.filter(f => f.supports(languageData));
```
Comment (manstis) [inline packages/microeditor-envelope/src/CompositeEditorFactory.ts]: I lookup a _delegate_ factory based on the _type_ of the `LanguageData`.

File: packages/microeditor-envelope/src/EditorFactory.ts [hunk 3]
```diff
@@ -22,6 +22,11 @@ import { EnvelopeBusInnerMessageHandler } from "./EnvelopeBusInnerMessageHandler
  * Factory of Editors to be created inside the envelope.
  */
 export interface EditorFactory<T extends LanguageData> {
+  /**
+   * Returns true if the factory supports a language.
+   * @param languageData The custom LanguageData
+   */
+  supports(languageData: T): boolean;
```
Comment (manstis) [inline packages/microeditor-envelope/src/EditorFactory.ts:28]: It was either to add this filtering of factory for type or have ALL factories attempt to build an editor and perform the filtering internally; returning an editor instance if they supported the type.. I thought this was cleaner.

File: packages/online-editor/src/App.tsx [hunk 4]
```diff
@@ -47,6 +48,7 @@ export function App(props: Props) {
   const router: EmbeddedEditorRouter = useMemo(
     () =>
       new EmbeddedEditorRouter(
+        new PMMLRoute(),
```
Comment (manstis) [inline packages/online-editor/src/App.tsx]: New route for new file type.

File: packages/online-editor/src/home/HomePage.tsx [hunk 5]
```diff
@@ -607,7 +607,7 @@ export function HomePage(props: Props) {
                 Choose a local file
                 {/* Transparent file input overlays the button */}
                 <input
-                  accept={".dmn, .bpmn, .bpmn2"}
+                  accept={".dmn, .bpmn, .bpmn2., .pmml"}
```
Comment (manstis) [inline packages/online-editor/src/home/HomePage.tsx]: Allow me to upload a local file to test.

File: packages/online-editor/static/images/pmml_kogito_logo.svg [hunk 6]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 23.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
+<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
```
Comment (manstis) [inline packages/online-editor/static/images/pmml_kogito_logo.svg]: Copy of DMN's SVG logo for the `EditorHeader`.. PMML shows as "DMN" for now....

File: packages/pmml-editor/src/editor/PMMLEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        return new Promise<void>(res => this.setState({ originalContent: content }, res)).then(() =>
+            this.updateContent(content)
+        );
+    }
+
+    private updateContent(content: string) {
+        return new Promise<void>(resolve => {
+            this.setState({ content: content }, () => {
+                console.log("Setting content to " + content);
+                this.props.messageBus.notify_dirtyIndicatorChange(this.isDirty());
```
Comment (manstis) [inline packages/pmml-editor/src/editor/PMMLEditor.tsx]: This mechanism seems obsolete. I'll need to integrate `StateControl`.. pointers welcome :-)

**Comment by manstis** [comment]
The last commit also moves a few bits around (only PMML stuff):-
- `pmml-editor-marshaller` - now only contains the marshaller (moved to a separate module as DMN has a need for just this marshaller to support PMML documents client-side; and trusty-ai needs it too to show a viewer for PMML documents - short term, will be moving to `kogito-tooling`).
- `pmml-editor` - the editor itself (stolen from [here](https://github.com/kiegroup/kogito-tooling-examples/tree/master/packages/simple-react-editors/src)!)

File: packages/pmml-editor/src/editor/PMMLEditor.tsx (hunk 7 above)
Comment (tiagobento) [reply packages/pmml-editor/src/editor/PMMLEditor.tsx]: Yes.. it is. I'll remove this message on a future PR.

**Comment by tiagobento** [comment]
@manstis Michael, can you please update this draft after #189 is merged? I guess it'll be so much easier for me to review :)

**Comment by manstis** [comment]
@tiagobento Absolutely... I have another PR in the pipeline first though (but not submitted until #189 is merged too)

**Comment by manstis** [comment]
@tiagobento This is ready for you to assassinate now :-)

The changes to `online-editor` are there to demonstrate _how_ additional types of editor _could_ be supported. I suspect the changes to `online-editor` themselves WILL NOT be merged in this PR (or at least the specifics for PMML will be removed). When you're happy with the approach I'll remove the PMML specific bits in `online-editor` and write tests where needed.

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts (hunk 2 above)
Comment (tiagobento) [reply packages/microeditor-envelope/src/CompositeEditorFactory.ts]: Nice! You don't need the explicit type declaration there though 😝

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ */
+export class CompositeEditorFactory implements EditorFactory<LanguageData> {
+  private factories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
+
+  constructor(factories: Array<EditorFactory<LanguageData>>) {
+    this.factories = factories;
+  }
+
+  public supports(languageData: LanguageData) {
+    return true;
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/CompositeEditorFactory.ts:31]: Shouldn't you check if the provided factories support the argument languageData too?

File: packages/microeditor-envelope/src/EditorFactory.ts (hunk 3 above)
Comment (tiagobento) [reply packages/microeditor-envelope/src/EditorFactory.ts:28]: Agreed.

File: packages/online-editor/src/home/HomePage.tsx (hunk 5 above)
Comment (tiagobento) [reply packages/online-editor/src/home/HomePage.tsx]: I'm assuming this is a temporary change so that you can test the PMML editor on the Online channel. Ideally we would have a rework on the Home screen if we decide to have the PMML editor on the Online channel as well, right?

File: packages/pmml-editor-marshaller/package.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    "build:fast": "rm -rf dist && webpack",
+    "build": "yarn run lint && yarn test && yarn run build:fast",
+    "build:prod": "yarn run build --mode production --devtool none"
+  },
+  "babel": {
+    "presets": [
+      "@babel/env"
+    ]
+  }
+}
```
Comment (tiagobento) [inline packages/pmml-editor-marshaller/package.json]: Thanks for that. I'm planning a rework on our package names and organisation and this really aligns well with it. I'll provide more details soon :D

File: packages/pmml-editor/src/editor/PMMLEditor.tsx (hunk 7 above)
Comment (tiagobento) [reply packages/pmml-editor/src/editor/PMMLEditor.tsx]: You'll want to update that once #200 is merged.

**Review by tiagobento — COMMENTED** [review]
Hi @manstis! I did a preliminary review of your code. I really like where this is going :)

In the coming weeks I'll try and reorganize our packages in a better way so that React-based editors can leverage a better API. For example, communicating with the channel using `window.envelope.*` is not really ideal.. It works very well for GWT-based editors, but we can do better for React.

#205 is the first step for that change, so if you wanna take a look there, I'm pretty sure you'll find it interesting :)

Let's keep in touch!

File: packages/pmml-editor-marshaller/package.json (hunk 9 above)
Comment (manstis) [reply packages/pmml-editor-marshaller/package.json]: ... and is essential for the bigger world!

- The DMN Editor in `kie-wb-common-dmn` needs a client-side PMML marshaller
- The Linear Regression Viewer in `trusty-ui-sandbox` also needs a PMML marshaller

The _plan_ is for both to reference this `pmml-marshaller` package.....

File: packages/online-editor/src/home/HomePage.tsx (hunk 5 above)
Comment (manstis) [reply packages/online-editor/src/home/HomePage.tsx]: Absolutely... ALL the changes in `online-editor` are just to showcase/demo the `CompositeEditorFactory`. When I say ALL, it's up for debate.. I _could_ keep the `CompositeEditorFactory` in the PR but don't register the `PMMLEditorFactory` or this change or the additional route...

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts (hunk 8 above)
Comment (manstis) [reply packages/microeditor-envelope/src/CompositeEditorFactory.ts:31]: Updated.

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts (hunk 2 above)
Comment (manstis) [reply packages/microeditor-envelope/src/CompositeEditorFactory.ts]: OK.. updated but I shiver at the concept (as I am a Java/typed-language developer)!

Comment (manstis) [reply packages/microeditor-envelope/src/CompositeEditorFactory.ts]: We'll need to think about what happens IF a factory cannot be found for a given `LanguageData`... we know it's not possible, but we don't protect against it (in the old AppFormer days we'd show a popup that an _activity could not be found_ -- it was loosely modelled on Android's _activities_).

**Comment by manstis** [comment]
Hi @tiagobento what/where/how etc are the changes you're making going to affect this?

I could do with the `pmml` package split (`pmml-marshaller`, `pmml-editor`) in `kogito-tooling` soon as the marshaller is needed elsewhere. If you think the impact on the changes I've made for React-based editors are going to be extensive I might as well split this PR into two; one adding the split for `pmml` packages - to be merged sooner that later - and the other PR worrying about a basic PMML editor implementation (wired into the `StateControl`, the new `Promise` work etc).

If you think the impact here is manageable and coming soon I'll sit tight...

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts (hunk 2 above)
Comment (tiagobento) [reply packages/microeditor-envelope/src/CompositeEditorFactory.ts]: Yeah.. that sounds like a good feature for the Envelope itself. Showing a nice message saying that something went wrong or that the editor is not available for that specific file extension.

File: packages/embedded-editor/src/envelope/envelope.ts (hunk 1 above)
Comment (tiagobento) [reply packages/embedded-editor/src/envelope/envelope.ts]: Yeah, I think that's the default configuration for TSLint.

File: packages/online-editor/src/home/HomePage.tsx (hunk 5 above)
Comment (tiagobento) [reply packages/online-editor/src/home/HomePage.tsx]: I like that the `CompositeEditorFactory` is in microeditor-envelope. People creating envelopes *will* have access to all the modules providing editors (along with their respective factories), so having the `CompositeEditorFactory` class there is nice.

File: packages/pmml-editor-marshaller/package.json (hunk 9 above)
Comment (tiagobento) [reply packages/pmml-editor-marshaller/package.json]: Nice! :)

**Comment by manstis** [comment]
@tiagobento Hello, this is "good to go" now...

I've removed the example use in `online-editor` and added unit tests.

I expect this PR to change before it gets to the front of the queue due to other PRs in the pipeline. 

@jomarko Given the above comment; feel free to leave any review for now, until @tiagobento has approved, other PRs in front of this in the queue have been merged and any fallout changes completed here.

File: packages/embedded-editor/src/envelope/envelope.ts [hunk 10]
```diff
@@ -14,19 +14,24 @@
...
-import { ChannelType, getOperatingSystem } from "@kogito-tooling/core-api";
+import { ChannelType, getOperatingSystem, LanguageData } from "@kogito-tooling/core-api";
 import { GwtEditorWrapperFactory } from "@kogito-tooling/kie-bc-editors";
 import * as MicroEditorEnvelope from "@kogito-tooling/microeditor-envelope";
+import { CompositeEditorFactory, EditorFactory } from "@kogito-tooling/microeditor-envelope";
 import { EnvelopeBusMessage } from "@kogito-tooling/microeditor-envelope-protocol";
 
+const editorFactories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
+editorFactories.push(new GwtEditorWrapperFactory());
+const compositeEditorFactory: CompositeEditorFactory = new CompositeEditorFactory(editorFactories);
```
Comment (tiagobento) [inline packages/embedded-editor/src/envelope/envelope.ts]:
```suggestion
const compositeEditorFactory = new CompositeEditorFactory([new GwtEditorWrapperFactory()]);
```

File: packages/embedded-editor/src/envelope/envelope.ts [hunk 11]
```diff
@@ -14,19 +14,24 @@
...
+
 MicroEditorEnvelope.init({
   container: document.getElementById("envelope-app")!,
   busApi: {
     postMessage<T>(message: EnvelopeBusMessage<T>, targetOrigin?: string, _?: any) {
       window.parent.postMessage(message, "*", _);
     }
   },
-  editorFactory: new GwtEditorWrapperFactory(),
+  editorFactory: compositeEditorFactory,
```
Comment (tiagobento) [inline packages/embedded-editor/src/envelope/envelope.ts]: You can even inline this variable here.

File: packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts [hunk 12]
```diff
@@ -32,13 +32,19 @@ declare global {
   }
 }
 
+export const FACTORY_TYPE: string = "gwt";
```
Comment (tiagobento) [inline packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]: You can drop the explicit type definition.

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Composite Factory of Editors to be created inside the envelope. This implementation delegates potential construction
+ * to specific concrete factories. The first concrete factory to return a promise to create an editor is selected.
+ */
+export class CompositeEditorFactory implements EditorFactory<LanguageData> {
+  private factories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
+
+  constructor(factories: Array<EditorFactory<LanguageData>>) {
+    this.factories = factories;
+  }
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/CompositeEditorFactory.ts]:
```suggestion
  constructor(private readonly factories: Array<EditorFactory<LanguageData>>) {
  }
```

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  public supports(languageData: LanguageData) {
+    return this.factories.filter(f => f.supports(languageData)).length > 0;
+  }
+
+  public createEditor(
+    languageData: LanguageData,
+    messageBus: EnvelopeBusInnerMessageHandler
+  ): Promise<AppFormer.Editor> {
+    const candidates = this.factories.filter(f => f.supports(languageData));
+    return candidates[0].createEditor(languageData, messageBus);
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/CompositeEditorFactory.ts]: Can you throw an error if `candidates.length !== 1`?

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    this.factories = factories;
+  }
+
+  public supports(languageData: LanguageData) {
+    return this.factories.filter(f => f.supports(languageData)).length > 0;
+  }
+
+  public createEditor(
+    languageData: LanguageData,
+    messageBus: EnvelopeBusInnerMessageHandler
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/CompositeEditorFactory.ts]: The signature of this method changed a bit on #205, just a heads up since it's before in the merge queue.

File: packages/pmml-editor/src/editor/PMMLEditor.tsx (hunk 7 above)
Comment (tiagobento) [reply packages/pmml-editor/src/editor/PMMLEditor.tsx]: I think the whole callback is not necessary anymore.

File: packages/pmml-editor/src/editor/PMMLEditor.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    //saving triggers this method, so we also update the originalContent by calling `this.setContent`
+    public getContent(): Promise<string> {
+        return this.setContent(this.state.path, this.state.content).then(() => this.state.content);
+    }
+
+    public isDirty() {
+        return this.state.content !== this.state.originalContent;
+    }
```
Comment (tiagobento) [inline packages/pmml-editor/src/editor/PMMLEditor.tsx]: This method can be deleted.

File: packages/pmml-editor/src/editor/PMMLEditorInterface.tsx [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        return this.self.getContent();
+    }
+
+    public getPreview(): Promise<string | undefined> {
+        return Promise.resolve(undefined);
+    }
+
+    public isDirty(): boolean {
+        return this.self.isDirty();
+    }
```
Comment (tiagobento) [inline packages/pmml-editor/src/editor/PMMLEditorInterface.tsx]: This can be removed.

File: packages/pmml-editor/src/editor/PMMLEditorInterface.tsx [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        return <PMMLEditor exposing={s => (this.self = s)} messageBus={this.messageBus} />;
+    }
+
+    public async undo(): Promise<void> {
+        //Place holder until StateControl is added for React based components.
+    }
+
+    public async redo(): Promise<void> {
+        //Place holder until StateControl is added for React based components.
+    }
```
Comment (tiagobento) [inline packages/pmml-editor/src/editor/PMMLEditorInterface.tsx:55]: Actually, that's it. You can create undo and redo methods on your PMMLEditor component that know how to undo and redo changes.

Comment (tiagobento) [reply packages/pmml-editor/src/editor/PMMLEditorInterface.tsx:55]: No need to change that in this PR, just FYI.

**Review by tiagobento — CHANGES_REQUESTED** [review]
Thanks for the PR, @manstis! Keep an eye on #205, since it will require you to update this PR once it's merged.

Some minor change requests inline :)

File: packages/embedded-editor/src/envelope/envelope.ts (hunk 10 above)
Comment (manstis) [reply packages/embedded-editor/src/envelope/envelope.ts]: Done and inlined.

File: packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts (hunk 12 above)
Comment (manstis) [reply packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]: Eh? Where did you see that?

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts (hunk 13 above)
Comment (manstis) [reply packages/microeditor-envelope/src/CompositeEditorFactory.ts]: Wow. That is some black-art voodoo syntax (that this ol' Java-boy will have to learn!)

File: packages/microeditor-envelope/src/CompositeEditorFactory.ts (hunk 14 above)
Comment (manstis) [reply packages/microeditor-envelope/src/CompositeEditorFactory.ts]: Sure thing done... and test added!

File: packages/pmml-editor/src/editor/PMMLEditorInterface.tsx (hunk 17 above)
Comment (manstis) [reply packages/pmml-editor/src/editor/PMMLEditorInterface.tsx]: Done.

File: packages/pmml-editor/src/editor/PMMLEditor.tsx (hunk 16 above)
Comment (manstis) [reply packages/pmml-editor/src/editor/PMMLEditor.tsx]: Done.

File: packages/pmml-editor/src/editor/PMMLEditorInterface.tsx (hunk 18 above)
Comment (manstis) [reply packages/pmml-editor/src/editor/PMMLEditorInterface.tsx:55]: OK, the `PMMLEditor` class is just a place holder at the moment. It was demo'ed briefly in the `online-editor` to prove different editor factories were being matched and used.. but since that code has been reverted this class now simple marks _hook_ for the real editor (that will live in `pmml-editor`...).

**Comment by manstis** [comment]
@tiagobento Updates made (other than [this](https://github.com/kiegroup/kogito-tooling/pull/197#discussion_r455902986) one as IDK where/what you exactly mean).

I'll be updating this after #205 so can make the change then... if you can point it out to me.

File: packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts (hunk 12 above)
Comment (tiagobento) [inline packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]:
```suggestion
export const FACTORY_TYPE = "gwt";
```

That's what I meant :D

**Comment by tiagobento** [comment]
@manstis Thanks! Just added a comment suggestion there :)

File: packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts (hunk 12 above)
Comment (manstis) [reply packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]: Oeeeee... well.... I _like_ using TS's type system :-)

Should we not be trying to use it more extensively? Is that not one of the reasons for using TS (over JS)?

Is this where JS vs Java developers worlds collide!! I hated writing JS and having to guess parameters types or stepping through implementations to find whether it was expected to be a primitive or object or either when documentation was missing...

Comment (tiagobento) [reply packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]: Yeah..... well, Java has `var` now too, which does basically the same type-inference mechanism :P

In TypeScript, string literals are types too, and since you're creating a `const` variable, the inferred type of `FACTORY_TYPE` is actually `"gwt"`, and not `string`. Sometimes I end up writing the types explicitly too, but only when I need to make something not ambiguous. 

Specially when I'm creating object literals, sometimes you want them to be of a specific type/interface, instead of the inferred. Interesting is that TypeScript matches types "by value", so two separate types/interfaces defining the exactly same properties are actually compatible.

One thing that helps, is on your IDE, you can probably hover over a variable/function to see its explicitly-typed signature. On WebStorm I get this:

image

This is more like a "cultural shock" between Java and TypeScript worlds, but TypeScript actually does a pretty decent job with its type-inference. I know it might look weird to you, but I believe this kind of stuff makes your code more idiomatic to other TypeScript developers and it's something you'll get used to read and write with the time :)

I don't think this is a requirement for me to approve this nor future PRs, since the code is actually correct and there are no major impacts. I know it's a lot to take in moving from Java to TypeScript, but I'd really like for us to try and "dive in" the TypeScript-way of writing code. Hope that makes sense you!

Comment (manstis) [reply packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]: OK, good to know.. I want to follow the _best practice_ for TS development.

Updated: https://github.com/kiegroup/kogito-tooling/pull/197/commits/2e14fcc651c8b5391adc1bc332ff35de95557365

I refer to Typescript's type/interface matching as "shape"... if the two JSON objects have the same structure they're the same... it works to our advantage nicely many times.

Comment (tiagobento) [reply packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]: Yes! Agreed :)

File: packages/online-editor/src/App.tsx [hunk 19]
```diff
@@ -23,6 +23,7 @@ import * as React from "react";
 import { useCallback, useMemo, useState } from "react";
 import { Route, Switch } from "react-router";
 import { HashRouter } from "react-router-dom";
+import "../static/resources/style.css";
```
Comment (tiagobento) [inline packages/online-editor/src/App.tsx]: Last one: Can you revert the changes on this file? We've had problems with CSS-loading order before and before we investigate that in more detail I don't wanna run any risks.

**Review by tiagobento — CHANGES_REQUESTED** [review]
Just one more comment I just saw. Ready to approve after that :)

File: packages/online-editor/src/App.tsx (hunk 19 above)
Comment (manstis) [reply packages/online-editor/src/App.tsx]: Done.

File: packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts (hunk 12 above)
Comment (manstis) [reply packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts]: Done.

**Review by tiagobento — APPROVED** [review]

**Comment by jomarko** [comment]
@manstis , hi please let me know once this has conflicts resolved

**Comment by manstis** [comment]
@jomarko @ederign This is ready for your review.

@tiagobento IDK if you want to re-review as some changes were needed after rebasing with `master`.

**Comment by tiagobento** [comment]
Hi @manstis, thanks for updating this PR. Can you please check your `webpack.config.js` files? I've created `common` file so that modules can define shorter configuration files. You can update your to use that.

**Comment by manstis** [comment]
@tiagobento I've updated the `webpack.config.js` file in a new commit.

**Review by tiagobento — APPROVED** [review]

**Comment by ederign** [comment]
@jomarko do you have any estimate when you will gonna be able to review this? This is blocking a bunch of other PR's.

**Review by ederign — APPROVED** [review]

**Comment by manstis** [comment]
@ederign @jomarko is on PTO.
